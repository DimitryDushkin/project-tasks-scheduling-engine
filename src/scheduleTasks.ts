import {
  makeGraphFromTasks,
  dfs,
  removeCyclicDependencies,
  toposort,
} from "./graph.utils";
import {
  addDays,
  getNowDate,
  maxDateTime,
  subtractDays,
  updateTaskDatesByStart,
} from "./date.utils";

export type ID = string;

export type Graph = Map<ID, Set<ID>>;
export type TasksById = { [id: string]: Task };
export type Task = {
  id: ID;
  title: string;
  start: Date;
  end: Date;
  duration: number;

  /**
   * Approximation of priority
   */
  position: number;
  progress: number;
  resourceId: ID;
  /**
   * Current task blocked by these tasks (depends on)
   */
  blockedBy?: Array<ID>;
};

export const scheduleTasks = (
  inputTasks: Array<Task>,
  today: Date = getNowDate()
): Array<Task> => {
  const dayBeforeToday = subtractDays(today, 1);
  const tasks: Array<Task> = inputTasks.map((t) => ({ ...t }));
  const tasksById: TasksById = Object.fromEntries(tasks.map((t) => [t.id, t]));
  const graph = makeGraphFromTasks(tasks);
  let cyclesToFullyUpdateDates = 1;

  // 1. Remove cyclic dependencies
  removeCyclicDependencies(graph, tasks);

  // 2. Initial update of all tasks start and ends days taking into account business days
  for (const task of tasks) {
    updateTaskDatesByStart(task, today, true);
  }

  // Repeat until dates remains unchanged, max graph.size times.
  // Similar to optimization in Bellman-Ford algorithm
  // @see https://en.wikipedia.org/wiki/Bellmanâ€“Ford_algorithm#Improvements
  for (let i = 0; i < tasks.length; i++) {
    let isAnyTaskTimelineChanged = false;

    for (const [taskId] of dfs(graph)) {
      const task = tasksById[taskId];
      // if blockedBy task not in initial data set
      if (task === undefined) {
        continue;
      }
      const blockedByTasks = Array.from(graph.get(task.id) ?? [])
        .map((blockedById) => tasksById[blockedById])
        // do not take into account tasks not in graph
        .filter(Boolean);
      const blockedByEndDates = blockedByTasks.map((t) => t.end);
      // add dayBeforeToday by default, so task without blockedBy starts on today
      blockedByEndDates.push(dayBeforeToday);

      // Start task on the next day after previous (blockedBy) tasks ends
      const maxBlockedByEndDate = addDays(maxDateTime(blockedByEndDates), 1);
      const isTaskTimelineUpdated = updateTaskDatesByStart(
        task,
        maxBlockedByEndDate
      );
      if (isTaskTimelineUpdated) {
        isAnyTaskTimelineChanged = true;
      }
    }

    if (isAnyTaskTimelineChanged === false) {
      break;
    }
    cyclesToFullyUpdateDates++;

    if (isAnyTaskTimelineChanged && i === tasks.length - 1) {
      console.error(
        'It\'s not enought "tasks.length" interations to fully schedule all tasks!'
      );
    }
  }

  console.debug(
    `Cycles to fully update dates ${cyclesToFullyUpdateDates}/${tasks.length}`
  );

  return toposort(graph, tasks);
};
