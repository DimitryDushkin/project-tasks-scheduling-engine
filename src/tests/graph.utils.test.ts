import {
  makeGraphFromTasks,
  makeReverseGraph,
  dfs,
  dfsWithLevels,
  toposort,
} from "../graph.utils";
import { Graph } from "../scheduleTasks";
import { tasks, tasksSingleResource } from "./mocks/tasks.mocks";

describe("graph.utils", () => {
  it("should make correct graph from tasks", () => {
    expect(makeGraphFromTasks(tasks)).toMatchSnapshot();
  });

  it("should make correct graph from tasks with single resource", () => {
    expect(makeGraphFromTasks(tasksSingleResource)).toMatchSnapshot();
  });

  it("should make correct reverse graph", () => {
    const graph: Graph = new Map([
      ["0", new Set(["1", "3"])],
      ["1", new Set("2")],
      ["2", new Set("3")],
    ]);

    expect(makeReverseGraph(graph)).toMatchSnapshot();
  });

  it("should correctly iterate over all nodes DFS", () => {
    const graph: Graph = new Map([
      ["0", new Set(["1", "3"])],
      ["1", new Set("2")],
      ["2", new Set("3")],
      ["3", new Set()],
      ["4", new Set()],
    ]);

    const nodes = {
      "0": 0,
      "1": 1,
      "2": 2,
      "3": 3,
      "4": 4,
    } as any;

    const expectedSum = 10;

    let resultSum = 0;

    for (const [id] of dfs(graph)) {
      if (id in nodes) {
        resultSum += nodes[id];
      }
    }
    expect(resultSum).toEqual(expectedSum);
  });

  it("should correctly iterate over all nodes DFS with nesting levels", () => {
    /**
          ┌───┐        ┌───┐
       ┌──┤ 0 ├───┐    │ 4 │
       │  └───┘   │    └───┘
     ┌─▼─┐     ┌──▼┐
     │ 1 │     │ 2 │
     └─┬─┘     └───┘
       │
     ┌─▼─┐
     │ 3 │
     └───┘
     */
    const graph: Graph = new Map([
      ["0", new Set(["1", "2"])],
      ["1", new Set("3")],
      ["2", new Set()],
      ["3", new Set()],
      ["4", new Set()],
    ]);

    const expected = {
      "0": [1, 1],
      "1": [2, 1.01],
      "2": [2, 1.02],
      "3": [3, 1.011],
      "4": [1, 2],
    } as any;

    for (const [id, level, position] of dfsWithLevels(graph)) {
      const [expectedLevel, expectedPosition] = expected[id];
      expect(level).toEqual(expectedLevel);
      expect(position).toEqual(expectedPosition);
    }
  });

  // @TODO:!
  it("should remove cyclic dependencies", () => {});

  it("makes correct topological sort shuffle", () => {
    const shuffledTasks = [tasks[3], tasks[0], tasks[2], tasks[1]];
    const shuffledTasks2 = [tasks[1], tasks[2], tasks[0], tasks[3]];
    const graph = makeGraphFromTasks(shuffledTasks);
    const sorted = toposort(graph, shuffledTasks);
    const sorted2 = toposort(graph, shuffledTasks2);
    expect(sorted.map((t) => t.id)).toMatchSnapshot();
    expect(sorted2.map((t) => t.id)).toMatchSnapshot();
  });
});
