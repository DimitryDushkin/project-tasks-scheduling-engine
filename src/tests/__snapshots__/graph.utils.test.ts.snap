// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph.utils makes correct topological sort shuffle 1`] = `
Array [
  "0",
  "1",
  "2",
  "3",
]
`;

exports[`graph.utils makes correct topological sort shuffle 2`] = `
Array [
  "0",
  "1",
  "2",
  "3",
]
`;

exports[`graph.utils should make correct graph from tasks 1`] = `
Map {
  "0" => Set {},
  "1" => Set {
    "0",
  },
  "2" => Set {},
  "3" => Set {
    "1",
    "2",
  },
}
`;

exports[`graph.utils should make correct graph from tasks with single resource 1`] = `
Map {
  "0" => Set {},
  "1" => Set {
    "0",
  },
  "2" => Set {
    "1",
  },
  "3" => Set {
    "1",
    "2",
  },
}
`;

exports[`graph.utils should make correct reverse graph 1`] = `
Map {
  "0" => Set {},
  "1" => Set {
    "0",
  },
  "3" => Set {
    "0",
    "2",
  },
  "2" => Set {
    "1",
  },
}
`;
